File: 1032. Stream of Characters.py
class TrieNode:
    def __init__(self):
        self.child = [None] * 26
        self.isend = False

class Trie:
    def __init__(self):
        self.root = None
    
    def add(self, word):
        if not self.root:
            self.root = TrieNode()
        p = self.root
        for c in word:
            if not p.child[ord(c)-ord('a')]:
                p.child[ord(c)-ord('a')] = TrieNode()
            p = p.child[ord(c)-ord('a')]
        p.isend = True
        

class StreamChecker:
    '''
    Thinking:
    1. build reverse trie to make it easy to traverse with stream
    2. keep a truncated stream of max word len to store all possible situations
    3. check each letter whether the reversed stream in trie, if one of them in trie, return True else return Flase
    '''
    def __init__(self, words: List[str]):
        self.n = max([len(w) for w in words])
        self.T = Trie()
        for w in words:
            self.T.add(w[::-1])
        self.s = []

    def query(self, letter: str) -> bool:
        self.s.append(letter)
        if len(self.s) > self.n:
            self.s.pop(0)
        node = self.T.root
        for c in self.s[::-1]:
            node = node.child[ord(c)-ord('a')]
            if not node:
                return False
            if node and node.isend:
                return True
        return False
        
        
# Your StreamChecker object will be instantiated and called as such:
# obj = StreamChecker(words)
# param_1 = obj.query(letter)

File: 1032. Stream of Characters.java
/**
 * No reverse solution
 */
class StreamChecker {
    public class TrieNode {
        TrieNode[] child;
        boolean isEnd;
        TrieNode() {
            child = new TrieNode[26];
            isEnd = false;
        }
    }
    
    TrieNode root;
    Queue<TrieNode> q;
    
    void add(String word) {
        int len = word.length();
        TrieNode n = root;
        for (int i=0; i<len; i++) {
            char c = word.charAt(i);
            if (n.child[c-'a'] == null) {
                n.child[c-'a'] = new TrieNode();
            } 
            n = n.child[c-'a'];
        }
        n.isEnd = true;
    }

    public StreamChecker(String[] words) {
        root = new TrieNode();
        q = new ArrayDeque<>();
        for (String word : words) {
            add(word);
        }
        q.offer(root);
    }
    
    public boolean query(char letter) {
        int len = q.size();
        boolean res = false;
        for (int i=0; i<len; i++) {
            TrieNode n = q.poll();
            if (n.child[letter-'a'] != null) {
                q.offer(n.child[letter-'a']);
                if (n.child[letter-'a'].isEnd) {
                    res = true;
                }
            }
        }
        q.offer(root);
        return res;
    }
}

class StreamChecker {
    
    class TrieNode {
        boolean isWord;
        TrieNode[] next = new TrieNode[26];
    }

    TrieNode root = new TrieNode();
    StringBuilder sb = new StringBuilder();

    public StreamChecker(String[] words) {
        createTrie(words);
    }

    public boolean query(char letter) {
        sb.append(letter);
        TrieNode node = root;
        for (int i = sb.length() - 1; i >= 0 && node != null; i--) {
            char c = sb.charAt(i);
            node = node.next[c - 'a'];
            if (node != null && node.isWord) {
                return true;
            }
        }
        return false;
    }

    private void createTrie(String[] words) {
        for (String s : words) {
            TrieNode node = root;
            int len = s.length();
            for (int i = len - 1; i >= 0; i--) {
                char c = s.charAt(i);
                if (node.next[c - 'a'] == null) {
                    node.next[c - 'a'] = new TrieNode();
                }
                node = node.next[c - 'a'];
            }
            node.isWord = true;
        }
    }
}

/**
 * Your StreamChecker object will be instantiated and called as such:
 * StreamChecker obj = new StreamChecker(words);
 * boolean param_1 = obj.query(letter);
 */